src/main.rs

```diff
@@ -1,69 +1,41 @@
 // SPDX-FileCopyrightText: 2025 blinry <mail@blinry.org>
 //
 // SPDX-License-Identifier: GPL-3.0-or-later
+mod process_tree;
 
-use std::path::PathBuf;
+use std::{
+    io::{stdout, Write},
+    path::PathBuf,
+};
 
-#[derive(Debug)]
-struct Process {
-    cwd: PathBuf,
-    tty: i32,
-}
-
-#[derive(Debug)]
-struct ProcessTree {
-    node: Process,
-    children: Vec<ProcessTree>,
-}
-
-impl ProcessTree {
-    fn new(pid: u32) -> ProcessTree {
-        let p = procfs::process::Process::new(pid as i32).expect("Process should exist");
-        let node = Process {
-            cwd: p.cwd().expect("Process should have a cwd"),
-            tty: p.stat().expect("Process should hav status info").tty_nr,
-        };
-
-        let t = p
-            .task_main_thread()
-            .expect("Process should have main thread");
-        let c_pids = t.children().expect("Task should have children");
-        let children = c_pids
-            .iter()
-            .map(|&c_pid| ProcessTree::new(c_pid))
-            .collect();
-        ProcessTree { node, children }
-    }
-
-    fn leaf_nodes_with_tty(&self) -> Vec<&Process> {
-        let leaves: Vec<&Process> = self
-            .children
-            .iter()
-            .flat_map(|c| c.leaf_nodes_with_tty())
-            .collect();
-
-        if !leaves.is_empty() {
-            return leaves;
-        }
-
-        if self.node.tty != 0 {
-            vec![&self.node]
-        } else {
-            vec![]
-        }
-    }
-}
+use process_tree::ProcessTree;
+use procfs::ProcResult;
 
-fn main() {
+fn get_path() -> ProcResult<PathBuf> {
     let t = ProcessTree::new(
         std::env::args()
             .nth(1)
-            .expect("PID should be provided as first argument")
-            .parse()
-            .expect("PID should be a number"),
-    );
-    let leaves = t.leaf_nodes_with_tty();
+            .ok_or("First argument is required.")?
```

Comment:
For consistency, I'd suggest dropping the "." in the message. (Because panic sometimes prints more details after messages like this, separated by a ":".)